drivers/spike/hub/speaker.c

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+/*
+ * API for the hub built-in button.
+ *
+ * Copyright (c) 2022 Embedded and Real-Time Systems Laboratory,
+ *                    Graduate School of Information Science, Nagoya Univ., JAPAN
+ */
+
+#include <kernel.h>
+#include <spike/hub/speaker.h>
+#include <pbdrv/sound.h>
+#include <stdint.h>
+#include <math.h>
+
+#define WAVEFORM_LENGTH 128
+static uint16_t waveform_data[WAVEFORM_LENGTH];
+static uint8_t  waveform_volume = 0;
+
+void hub_speaker_set_volume(uint8_t volume)
+{
+  if (volume == waveform_volume) return;
```

Comment:
waveform_volumeをキャッシュのような役割をもたせるとしたら，以下が必要になりますね．
```suggestion
  if (volume == waveform_volume) return;
  waveform_volume = volume;
```